internal/controller/controllers/cluster_controller.go

```diff
@@ -92,7 +122,8 @@ func (r *ClusterReconciler) createNewCluster(ctx context.Context, cluster *adiio
 
 	pullSecret, err := r.getPullSecret(ctx, spec.PullSecretRef.Name, spec.PullSecretRef.Namespace)
 	if err != nil {
-		return ctrl.Result{}, errors.Wrap(err, "failed to get pull secret")
+		r.Log.WithError(err).Error("failed to get pull secret")
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
Instead of requeueing, which could produce a lot of churn or even a hot loop, it is better to Watch Secrets in your namespace, and cause a reconcile to happen for the cluster when its secret has an event. That way if the user creates the secret, or modifies it, that will implicitly cause Reconcile to run.

In SetupWithManager, you'll need to tell it to watch Secrets with this handler: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/handler#EnqueueRequestsFromMapFunc

Then in the map func, you'll need to write the logic of how it knows which cluster corresponds to the Secret that had the event. It looks like from this code that you're just assuming the secret has the same name and namespace as the cluster.

The example in that doc does not quite match the "builder" style you see in this scaffolded code, but you'll just use the [Watches](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/builder#Builder.Watches) function.

Reply:
I agree, i didn't want to make a big PR and we have a ticket to enable it for images.
Actually when i'm thinking about it, we have a bug here - we require pull secret for the Image resource but because rest api require it for the cluster level in CRDs we have both, now if we have different pull secret for the cluster and for the image then we will have a race and each resource will try to update it.
@avishayt maybe we need to remove the pull secret from the image resource for now?

Reply:
Regarding the pull-secret, it will remain only in the cluster resource, because we are using it both for the image and for another API the get the ignition files during the installation.

Reply:
I added a logic that watch for secret changes, what about he return value in this case? if i failed to get the secret what should be the return value?